pallets/msa/src/lib.rs

```diff
@@ -853,6 +858,51 @@ pub mod pallet {
 			}
 			Ok(Some(T::WeightInfo::retire_msa(num_deletions)).into())
 		}
+
+		/// create a provider by means of governance approval
+		#[pallet::call_index(11)]
+		#[pallet::weight(T::WeightInfo::create_provider(100_000))]
+		pub fn create_provider_via_governance(
+			origin: OriginFor<T>,
+			create_provider_proof: Signature,
+			create_provider_payload: ProviderData<T>,
+		) -> DispatchResult {
+			let _ = T::CreateProviderOrigin::ensure_origin(origin)?;
+
+			let create_provider_key = create_provider_payload.clone().provider_key;
+
+			// TODO: new Error message
+			Self::verify_signature(
+				&create_provider_proof,
+				&create_provider_key,
+				create_provider_payload.encode(),
+			)
+			.map_err(|_| Error::<T>::NewKeyOwnershipInvalidSignature)?;
+
+			let provider_msa_id = Self::ensure_valid_msa_key(&create_provider_key)?;
+
+			let bounded_name: BoundedVec<u8, T::MaxProviderNameSize> = create_provider_payload
+				.clone()
+				.provider_name
+				.try_into()
+				.map_err(|_| Error::<T>::ExceedsMaxProviderNameSize)?;
+
```

Comment:
Is there a way we could share the same code for this function and `create_provider()` so that changes only need to happen in one place?